lib/netext/httpext/request_test.go

```diff
@@ -357,3 +359,116 @@ func TestTrailFailed(t *testing.T) {
+	t.Parallel()
+	ln, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatal(err)
+	}
+	addr := ln.Addr()
+	defer func() {
+		require.NoError(t, ln.Close())
+	}()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	samples := make(chan stats.SampleContainer, 10)
+	logger := logrus.New()
+	logger.Level = logrus.DebugLevel
+	state := &lib.State{
+		Options: lib.Options{
+			RunTags:    &stats.SampleTags{},
+			SystemTags: &stats.DefaultSystemTagSet,
+		},
+		Transport: &http.Transport{
+			DialContext: (&net.Dialer{
+				Timeout: 1 * time.Microsecond,
+			}).DialContext,
+		},
+		Samples: samples,
+		Logger:  logger,
+		BPool:   bpool.NewBufferPool(100),
+	}
+
+	ctx = lib.WithState(ctx, state)
+	req, _ := http.NewRequest("GET", "http://"+addr.String(), nil)
+	preq := &ParsedHTTPRequest{
+		Req:              req,
+		URL:              &URL{u: req.URL, URL: req.URL.String()},
+		Body:             new(bytes.Buffer),
+		Timeout:          500 * time.Millisecond,
+		ResponseCallback: func(i int) bool { return i == 0 },
+	}
+
+	res, err := MakeRequest(ctx, preq)
+	require.NoError(t, err)
+	assert.NotNil(t, res)
+	assert.Len(t, samples, 1)
+	sampleCont := <-samples
+	allSamples := sampleCont.GetSamples()
+	require.Len(t, allSamples, 9)
+	expTags := map[string]string{
+		"error":             "dial: i/o timeout",
+		"error_code":        "1211",
+		"status":            "0",
+		"expected_response": "true", // we wait for status code 0
+		"method":            "GET",
+		"url":               req.URL.String(),
+		"name":              req.URL.String(),
+	}
+	for _, s := range allSamples {
+		assert.Equal(t, expTags, s.Tags.CloneTags())
+	}
+}
+
+func TestMakeRequestTimeoutInTheBegining(t *testing.T) {
+	t.Parallel()
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(100 * time.Millisecond)
+	}))
+	defer srv.Close()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	samples := make(chan stats.SampleContainer, 10)
+	logger := logrus.New()
+	logger.Level = logrus.DebugLevel
+	state := &lib.State{
+		Options: lib.Options{
+			RunTags:    &stats.SampleTags{},
+			SystemTags: &stats.DefaultSystemTagSet,
+		},
+		Transport: srv.Client().Transport,
+		Samples:   samples,
+		Logger:    logger,
+		BPool:     bpool.NewBufferPool(100),
+	}
+	ctx = lib.WithState(ctx, state)
+	req, _ := http.NewRequest("GET", srv.URL, nil)
+	preq := &ParsedHTTPRequest{
+		Req:              req,
+		URL:              &URL{u: req.URL, URL: srv.URL},
+		Body:             new(bytes.Buffer),
+		Timeout:          50 * time.Millisecond,
+		ResponseCallback: func(i int) bool { return i == 0 },
+	}
+
+	res, err := MakeRequest(ctx, preq)
+	require.NoError(t, err)
+	assert.NotNil(t, res)
+	assert.Len(t, samples, 1)
+	sampleCont := <-samples
+	allSamples := sampleCont.GetSamples()
+	require.Len(t, allSamples, 9)
+	expTags := map[string]string{
+		"error":             "Request timeout",
```

Comment:
rebase on the latest `master` and fix the capitalization? otherwise it's going to fail when you merge the PR